sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoSchema.java

```diff
@@ -0,0 +1,365 @@
+        descriptor.getOneofs().stream()
+            .flatMap(oneofDescriptor -> oneofDescriptor.getFields().stream())
+            .collect(Collectors.toList());
+
+    ArrayList<ProtoRow.FieldOverlay> fieldOverlays = new ArrayList<>();
+    Iterator<Descriptors.FieldDescriptor> protoFields = descriptor.getFields().iterator();
+    for (int i = 0; i < descriptor.getFields().size(); i++) {
+      Descriptors.FieldDescriptor protoField = protoFields.next();
+      if (protoField.isRepeated() && !isMap(protoField)) {
+        fieldOverlays.add(new ProtoRow.ArrayOverlay(this, protoField));
+      } else {
+        fieldOverlays.add(createFieldLayer(protoField, oneofMap.contains(protoField)));
+      }
+    }
+    return fieldOverlays;
+  }
+
+  private void init() {
+    this.getters = createFieldLayer(descriptor);
+
+    Schema.Builder builder = Schema.builder();
+    for (ProtoRow.FieldOverlay field : getters) {
+      builder.addField(field.getSchemaField());
+    }
+
+    schema = builder.build();
+    schema.setUUID(UUID.randomUUID());
+    globalSchemaCache.put(schema.getUUID(), this);
+    try {
+      if (DynamicMessage.class.equals(rawType)) {
+        this.fnNewBuilder = rawType.getMethod("newBuilder", Descriptors.Descriptor.class);
+      } else {
+        this.fnNewBuilder = rawType.getMethod("newBuilder");
+      }
+    } catch (NoSuchMethodException e) {
+    }
+  }
+
+  public Schema getSchema() {
+    return this.schema;
+  }
+
+  SchemaCoder<Message> getSchemaCoder() {
+    return SchemaCoder.of(schema, getToRowFunction(), getFromRowFunction());
+  }
+
+  private SerializableFunction<Message, Row> getToRowFunction() {
+    return new MessageToRowFunction(this);
+  }
+
+  private SerializableFunction<Row, Message> getFromRowFunction() {
+    return new RowToMessageFunction(this);
+  }
+
+  private void writeObject(ObjectOutputStream oos) throws IOException {
+    oos.defaultWriteObject();
+    if (DynamicMessage.class.equals(this.rawType)) {
+      if (this.descriptor == null) {
+        throw new RuntimeException("DynamicMessages require provider a Descriptor to the coder.");
+      }
+      oos.writeUTF(descriptor.getFile().getName());
+      oos.writeUTF(descriptor.getName());
+    }
+  }
+
+  private void readObject(ObjectInputStream ois) throws ClassNotFoundException, IOException {
+    ois.defaultReadObject();
+    if (DynamicMessage.class.equals(rawType)) {
+      descriptor = domain.getDescriptor(ois.readUTF(), ois.readUTF());
+    } else {
+      descriptor = ProtobufUtil.getDescriptorForClass(rawType);
+    }
+    init();
+  }
+
+  public Map<String, Class> getRegisteredTypeMapping() {
+    return typeMapping;
+  }
+
+  public ProtoDomain getDomain() {
+    return domain;
+  }
+
+  /** Overlay. */
+  public static class ProtoOverlayFactory implements Factory<List<FieldValueGetter>> {
+
+    public ProtoOverlayFactory() {}
+
+    @Override
+    public List create(Class<?> clazz, Schema schema) {
```

Comment:
should this return List<FieldValueGetter?>

Reply:
yes, changed to include it.